controllers/admin/adminmarketingestep3.php

```diff
@@ -624,11 +665,35 @@ private function cleanHTML()
 		$this->html_content = $html_cleaner->cleanHTML($this->html_content);
 	}
 
+	private function encodeURLs()
+	{
+		$images = $this->getImgTags($this->html_content);
+		$srcs = array_unique($images[2]);
+		foreach ($srcs as $key => $src)
+		{
+			$corrected_src = implode('/', array_map('rawurlencode', explode('/', $src)));
+			$corrected_src = str_replace('%3A//', '://', $corrected_src);
+			$corrected_src = str_replace('%23%23', '##', $corrected_src);
+			$this->html_content = str_replace($images[0][$key], $images[1][$key].$corrected_src.$images[3][$key], $this->html_content);
+		}
+
+		$links = $this->getLinkTags($this->html_content);
+		$hrefs = array_unique($links[2]);
+		foreach ($hrefs as $key => $href)
+		{
+			$corrected_link = implode('/', array_map('rawurlencode', explode('/', $href)));
+			$corrected_link = str_replace('%3A//', '://', $corrected_link);
+			$corrected_link = str_replace('%23%23', '##', $corrected_link);
+			$this->html_content = str_replace($links[0][$key], $links[1][$key].$corrected_link.$links[3][$key], $this->html_content);
```

Comment:
You have a common code for your loops. You execute the same loop on a variable which can be huge. Here is a lighter version which seems to fit your needs (not tested):

```php
	private function encodeURLs()
	{
		$elements = array_merge($this->getImgTags($this->html_content), $this->getLinkTags($this->html_content));
		$srcs = array_unique($elements[2]);
		foreach ($srcs as $key => $src)
		{
			$corrected_src = implode('/', array_map('rawurlencode', explode('/', $src)));
			$corrected_src = str_replace(array('%3A//', '%23%23'), array('://', '##'), $corrected_src);
			$this->html_content = str_replace($elements[0][$key], $elements[1][$key].$corrected_src.$elements[3][$key], $this->html_content);
		}
	}
```

With this version, you spare 1 loop execution and 1 call to `str_replace`.

With this version, you should gain some execution time :)